File: smartlist/src/main/java/com/smartlist/module/question/dto/response/QuestionRespDTO.java
package com.smartlist.module.question.dto.response;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.smartlist.model.Question;
import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import lombok.Getter;
import lombok.Setter;

import java.util.ArrayList;
import java.util.List;

@Getter
@Setter
public class QuestionRespDTO {

    @Schema(description = "Question Id", example = "", requiredMode= Schema.RequiredMode.REQUIRED)
    @NotNull
    @Positive
    Integer id;

    @Schema(description = "Question type", example = "", requiredMode= Schema.RequiredMode.REQUIRED)
    @NotEmpty
    String type;

    @Schema(description = "Question text", example = "", requiredMode= Schema.RequiredMode.REQUIRED)
    @NotEmpty
    String text;

    @Schema(description = "Question score", example = "", requiredMode= Schema.RequiredMode.REQUIRED)
    @NotNull
    @Positive
    Integer score;

    @Schema(description = "Question sort order", example = "", requiredMode= Schema.RequiredMode.REQUIRED)
    @NotNull
    @Positive
    @JsonProperty(value = "sort_order")
    Integer sortOrder;

    @Schema(description = "Answer options", example = "", requiredMode= Schema.RequiredMode.REQUIRED)
    @NotEmpty
    @JsonProperty(value = "answer_options")
    List<AnswerOptionRespDTO> answerOptions;

    @Schema(description = "Question options", example = "", requiredMode= Schema.RequiredMode.REQUIRED)
    @NotEmpty
    String options;

    /*
    public QuestionRespDTO(Question question, boolean forAssessment) {
        this.id = question.getId();
        this.type = question.getType().toString();
        this.text = question.getText();
        this.score = question.getScore();
        this.options = question.getOptions();
        this.sortOrder = question.getSortOrder();
        this.answerOptions = new ArrayList<>();
        question.getAnswerOptions().forEach(answerOption -> this.answerOptions.add(new AnswerOptionRespDTO(answerOption, forAssessment)));
    }*/


}

File: javascript/smartlist/smartlist-api-mysql-project/src/modules/question/dto/response/question.resp.dto.ts
import { ApiProperty } from '@nestjs/swagger';
import { IsInt, IsNotEmpty } from 'class-validator';
import { AnswerOption } from 'src/entities/answer-option.entity';
import { Question } from 'src/entities/question.entity';
import { QuestionOptionReqDto } from '../request/options.req.dto';
import { AnswerOptionRespDto } from './answer-option.resp.dto';

export class QuestionRespDto {
  @ApiProperty({
    type: Number,
  })
  @IsNotEmpty()
  @IsInt()
  id: number;

  @ApiProperty({
    type: String,
  })
  @IsNotEmpty()
  type: string;

  @ApiProperty({
    type: String,
  })
  @IsNotEmpty()
  text: string;

  @ApiProperty({
    type: Number,
  })
  @IsNotEmpty()
  score: number;

  @ApiProperty({
    type: Number,
  })
  @IsNotEmpty()
  @IsInt()
  sort_order: number;

  @ApiProperty({
    type: [AnswerOptionRespDto],
  })
  @IsNotEmpty()
  answer_options: AnswerOptionRespDto[];

  @ApiProperty({
    type: JSON,
  })
  options: QuestionOptionReqDto;

  constructor(question: Question, forAssessment: boolean) {
    this.id = question.id;
    this.type = question.type;
    this.text = question.text;
    this.score = question.score;
    this.options = question.options;
    this.sort_order = question.sortOrder;
    this.answer_options = [];
    question.answerOptions.forEach((answerOption: AnswerOption) => {
      this.answer_options.push(
        new AnswerOptionRespDto(answerOption, forAssessment),
      );
    });
  }
}
